internal/keepers/service.go

```diff
@@ -131,7 +143,56 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, keys ...types.U
 
 	// Cache results
 	for i, u := range checkResults {
-		s.cache.Set(string(keys[nonCachedKeysIdxs[i]]), u, util.DefaultCacheExpiration)
+		s.checkCache.Set(string(keys[nonCachedKeysIdxs[i]]), u, util.DefaultCacheExpiration)
+		results[nonCachedKeysIdxs[i]] = u
+	}
+
+	return results, nil
+}
+
+func (s *onDemandUpkeepService) GetUpkeep(ctx context.Context, keys ...types.UpkeepKey) ([]types.UpkeepInfo, error) {
+	var (
+		wg                sync.WaitGroup
+		results           = make([]types.UpkeepInfo, len(keys))
+		nonCachedKeysLock sync.Mutex
+		nonCachedKeysIdxs = make([]int, 0, len(keys))
+		nonCachedKeys     = make([]types.UpkeepKey, 0, len(keys))
+	)
+
+	for i, key := range keys {
+		wg.Add(1)
+		go func(i int, key types.UpkeepKey) {
+			// the cache is a collection of keys (block & id) that map to cached
+			// results. if the same upkeep is checked at a block that has already been
+			// checked, return the cached result
+			if result, cached := s.getCache.Get(string(key)); cached {
+				results[i] = result
+			} else {
+				nonCachedKeysLock.Lock()
+				nonCachedKeysIdxs = append(nonCachedKeysIdxs, i)
+				nonCachedKeys = append(nonCachedKeys, key)
+				nonCachedKeysLock.Unlock()
+			}
+			wg.Done()
+		}(i, key)
+	}
+
+	wg.Wait()
+
+	// All keys are cached
+	if len(nonCachedKeys) == 0 {
+		return results, nil
+	}
+
+	// get upkeep info at block number in key
+	getResults, err := s.registry.GetUpkeep(ctx, nonCachedKeys...)
+	if err != nil {
+		return nil, fmt.Errorf("%w: service failed to get upkeep from registry", err)
+	}
+
+	// Cache results
+	for i, u := range getResults {
+		s.getCache.Set(string(keys[nonCachedKeysIdxs[i]]), u, util.DefaultCacheExpiration)
```

Comment:
Not sure if we should use cache here. This will result in more time a new gas limit takes to be reflected in our system

Reply:
Unless we listen for gas limit updated logs and clear the cache

Reply:
This is the cache per upkeep per block so there shouldn't be any issues with this.

Reply:
yup, cache per block should be fine